src/eth_rpc/servers/alchemy_rpc.rs

```diff
@@ -1,41 +1,83 @@
 use futures::future::join_all;
 use jsonrpsee::core::{async_trait, RpcResult};
-use reth_primitives::{Address, BlockId, BlockNumberOrTag};
+use reth_primitives::{Address, BlockId, BlockNumberOrTag, U256};
 
 use crate::eth_provider::contracts::erc20::EthereumErc20;
 use crate::eth_provider::error::EthApiError;
+use crate::eth_provider::provider::EthereumProvider;
 use crate::eth_rpc::api::alchemy_api::AlchemyApiServer;
-use crate::models::balance::TokenBalanceFuture;
-use crate::{eth_provider::provider::EthereumProvider, models::balance::TokenBalances};
+use crate::models::token::{TokenBalance, TokenBalances, TokenMetadata};
 
 /// The RPC module for the Ethereum protocol required by Kakarot.
 #[derive(Debug)]
 pub struct AlchemyRpc<P: EthereumProvider> {
+    /// The provider for interacting with the Ethereum network.
     eth_provider: P,
 }
 
 impl<P: EthereumProvider> AlchemyRpc<P> {
+    /// Creates a new instance of [`AlchemyRpc`].
     pub const fn new(eth_provider: P) -> Self {
         Self { eth_provider }
     }
 }
 
 #[async_trait]
 impl<P: EthereumProvider + Send + Sync + 'static> AlchemyApiServer for AlchemyRpc<P> {
+    /// Retrieves the token balances for a given address.
     #[tracing::instrument(skip_all, ret, fields(address = %address, token_addresses = ?token_addresses))]
     async fn token_balances(&self, address: Address, token_addresses: Vec<Address>) -> RpcResult<TokenBalances> {
         tracing::info!("Serving alchemy_getTokenBalances");
 
+        // Set the block ID to the latest block
         let block_id = BlockId::Number(BlockNumberOrTag::Latest);
-        let handles = token_addresses.into_iter().map(|token_addr| {
-            let token = EthereumErc20::new(token_addr, &self.eth_provider);
-            let balance = token.balance_of(address, block_id);
 
-            TokenBalanceFuture::new(Box::pin(balance), token_addr)
-        });
+        Ok(TokenBalances {
+            address,
+            token_balances: join_all(token_addresses.into_iter().map(|token_address| async move {
+                // Create a new instance of `EthereumErc20` for each token address
+                let token = EthereumErc20::new(token_address, &self.eth_provider);
+                // Retrieve the balance for the given address
+                let token_balance = token.balance_of(address, block_id).await?;
+                Ok(TokenBalance { token_address, token_balance })
+            }))
```

Comment:
here I've removed the old `TokenBalanceFuture` because with the new more structured implementation there were lifetime issues outside the map because of the old Box::pin. I believe that this way of doing things is much simpler and it does not require a special lifetime, it is the good use case for async move:

https://rust-lang.github.io/async-book/03_async_await/01_chapter.html#async-move

> async blocks and closures allow the move keyword, much like normal closures. An async move block will take ownership of the variables it references, allowing it to outlive the current scope, but giving up the ability to share those variables with other code:

Let me know what you think about this but basically this is the same

Reply:
makes a lot of sense yes!